sources/SilkTouch/Caching/FileSystemCacheProvider.cs

```diff
@@ -0,0 +1,243 @@
+        await (sema?.WaitAsync() ?? Task.CompletedTask);
+        if (
+            sema is not null
+            && ((isUnstaged && !Directory.Exists(path)) || (!isUnstaged && !File.Exists(path)))
+        )
+        {
+            if ((flags & CacheFlags.AllowNewLocked) == 0)
+            {
+                sema.Release();
+                return null;
+            }
+
+            logger.LogDebug("Cache miss: {} for {}", cacheKey, intent);
+            if (isUnstaged)
+            {
+                if (File.Exists(path))
+                {
+                    File.Delete(path);
+                }
+
+                Directory.CreateDirectory(path);
+            }
+            else
+            {
+                if (Directory.Exists(path))
+                {
+                    Directory.Delete(path, true);
+                }
+
+                if (Path.GetDirectoryName(path) is { } dn && !Directory.Exists(dn))
+                {
+                    Directory.CreateDirectory(dn);
+                }
+
+                if (File.Exists(path) && (flags & CacheFlags.NoHostDirectory) == 0)
+                {
+                    File.Move(path, path + ".tmp");
+                    {
+                        await using var fs = File.OpenRead(path);
+                        using var arch = new ZipArchive(fs);
+                        await Parallel.ForEachAsync(
+                            arch.Entries,
+                            async (entry, ct) =>
+                            {
+                                var options = new FileStreamOptions
+                                {
+                                    Access = FileAccess.Write,
+                                    Mode = FileMode.Create,
+                                    Share = FileShare.None,
+                                    BufferSize = 4096
+                                };
+                                var unixFileMode = (UnixFileMode)(
+                                    (entry.ExternalAttributes >> 16) & 511
+                                );
+                                if (
+                                    unixFileMode != UnixFileMode.None
+                                    && !OperatingSystem.IsWindows()
+                                )
+                                {
+                                    options.UnixCreateMode = unixFileMode;
+                                }
+
+                                var dstPath = Path.Combine(path, entry.FullName);
+                                Directory.CreateDirectory(Path.GetDirectoryName(dstPath) ?? path);
+                                await using var destination = new FileStream(dstPath, options);
+                                await using var stream = entry.Open();
+                                await stream.CopyToAsync(destination, ct);
+                                await destination.FlushAsync(ct);
```

Comment:
does disposeasync not flush?

Reply:
From what I can find, CopyToAsync does not Flush automatically, so this does seem correct

Reply:
its being await using'd though, my question is about DisposeAsync not the copyto

Reply:
I don't think dispose will necessarily flush either, but I'm unsure

Reply:
In my experience it’s not necessarily the case, and indeed before I added flushing for the _noHostZips I was getting corrupted archives, so I think it is technically required.

Reply:
alright, just wanted to check